src/freenet/io/comm/MessageFilter.java

```diff
@@ -46,8 +45,8 @@
     private boolean _matched;
     private PeerContext _droppedConnection;
     private MessageType _type;
-    private final HashMap<String, Object> _fields = new HashMap<String, Object>();
-    private final List<String> _fieldList = new LinkedList<String>();
+    private final List<Object> _fields = new ArrayList<Object>();
```

Comment:
I agree that this is not the cleanest solution, but `LinkedHashMap` is definitely the worst choice in terms of memory efficiency. Note that we actually don't care in which particular order the fields are matched.

If you feel clean code outweighs memory and computational efficiency here, I will replace the `ArrayList`s with a single `HashMap` and use the `entrySet()` to iterate over. Considering this list is traversed for every single packet we receive (with high probability), I'd rather keep the most lightweight approach.